api/ondemand_vhds_rds_protocol.rst

```diff
@@ -0,0 +1,111 @@
+.. _on_demand_vhds_rds_protocol:
+
+On-Demand RDS/VHDS
+==================
+Currently, in RDS, all routes for the cluster are sent to every Envoy instance
+in the mesh. This causes scaling issues as the size of the cluster grows. The
+majority of this complexity can be found in the virtual host configurations, of
+which most are not needed by any individual proxy. With a goal of being able
+to scale to one million vhosts in the future, the capability to filter a
+proxy to only contain the required virtual hosts is a necessity.
+
+In order to fix this issue, we are implementing the on-demand Virtual Host
+Discovery Service (VHDS). On-demand VHDS uses the delta xDS protocol, which is
+a separate protocol from xDS. Instead of sending all virtual hosts with a route
+config, using VHDS will allow an Envoy instance to subscribe and unsubscribe
+from a list of virtual hosts stored internally in the xDS management server.
+The xDS management server will monitor this list and use it to filter the
+configuration sent to an individual Envoy instance to only contain the
+subscribed virtual hosts.
+
+Joining/Reconnecting
+====================
+When an Envoy instance forms a delta xDS connection to the xDS management
+server for the first time, it will receive a base configuration filtered down
+to a subset of routes that are likely useful to the Envoy instance. For
+example, these might be filtered down by routes that exist in the current
+namespace. The xDS management server will send this same collection of
+resources on reconnect.
+
+.. figure:: diagrams/delta_rds_connection.svg
+   :alt: Delta RDS connection
+
+Requesting Additional resources
+===============================
+In order to request additional resources, Envoy will send a
+:ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>`
+to a xDS management server. The xDS management server will then respond with a
+:ref:`DeltaDiscoveryResponse <envoy_api_msg_DeltaDiscoveryResponse>`
+which is followed up by an Ack/Nack
+:ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>`
+from Envoy.
+
+.. figure:: diagrams/delta_rds_request_additional_resources.svg
+   :alt: Delta RDS request additional resources
+
+Virtual Host Discovery Service
+==============================
+In order to support on-demand, an additional protocol, VHDS, will be added.
+This protocol allows a separation of concerns with RDS where RDS is in charge
+of maintaining route configs while VHDS is in charge of communicating virtual
+hosts. In VHDS, Envoy will send a
+:ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>`
+with :ref:`type_url` set to 
+`type.googleapis.com/envoy.api.v2.route.VirtualHost` 
+and :ref:`resource_names <envoy_api_msg_DeltaDiscoveryRequest.resource_names>` 
+set to a list of routeconfig names + the request HTTP "host"
```

Comment:
Please specify the grammar for this more formally, this is a key part of the protocol.

Reply:
We still need to discuss the formal grammar. For now I'm using routeConfigName$host, but that's likely not the best solution.

Reply:
@dmitri-d can we agree on something here? I think we can take the ideas from UDPA-TP and just use the sub-resource naming scheme from there. WDYT?

Reply:
From the UDPA-TP spec, I see:
<namespace>/<resource> 

We need the route config name, so how about: 

<routeConfigName>/<resource> 

?

Reply:
Yep, this is good for now, as discussed on the call today, the resource will not have any slashes, so we're all good.